### ISSUE TITLE
M2.2.4 : Extension attributes can not be used for product attributes

### ISSUE THREAD

**Issue by kanduvisla** [magento/magento2#16406]
I'm trying to add extension attributes to product attributes, but I'm getting the error:

```
Method 'getExtensionAttributes' must be overridden in the interfaces which extend 'Magento\Framework\Api\ExtensibleDataInterface'. Concrete return type must be specified. Please fix :\Magento\Catalog\Api\Data\ProductAttributeInterface
```

### Preconditions

1. Magento 2.2.4

### Steps to reproduce

I declare my extension attributes in `etc/extension_attributes.xml`:

```
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="urn:magento:framework:Api/etc/extension_attributes.xsd">
    <extension_attributes for="Magento\Catalog\Api\Data\ProductAttributeInterface">
        <attribute code="attribute_suffix" type="string"/>
        <attribute code="trim_trailing_zeros" type="boolean"/>
    </extension_attributes>
</config>
```

I use the `catalog_entity_attribute_save_after`-event to set extension attributes:

```
$attribute = $observer->getEvent()->getData('attribute');

if ($attribute->getExtensionAttributes() === null) {
    /** \Magento\Framework\Api\ExtensionAttributesFactory */
    $extensionAttributes = $this->extensionAttributesFactory->create(ProductAttributeInterface::class);
    $attribute->setExtensionAttributes($extensionAttributes);
}
```

Please note that for some reason, the `ProductAttributeInterface` does have a `getExtensionAttributes()`-method, but not a `setExtensionAttributes()`, but the implementation (`\Magento\Catalog\Model\ResourceModel\Eav\Attribute`) inherits from `\Magento\Eav\Model\Entity\Attribute\AbstractAttribute` and this does have that method.

### Expected result

I should be able to add extension attributes for product attributes

### Actual result

I'm getting the error as described above.

I'm guessing that the lack of `get / setExtensionAttributes` in `ProductAttributeInterface` is responsible for this error. Is this a bug? Because the methods are there (even though provided through inheritance). Or are you simple not allowed to use extension attributes in this specific case?

**Comment by magento-engcom-team** [comment]
Hi @kanduvisla. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me {$VERSION} instance` 

where `{$VERSION}` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (2.2-develop +). 
For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation.

@kanduvisla do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by orlangur** [comment]
@kanduvisla https://devdocs.magento.com/guides/v2.2/extension-dev-guide/extension_attributes/adding-attributes.html looks like a working example of adding extension attributes to product entity.

**Comment by kanduvisla** [comment]
@orlangur : that's right, and the example is about add extension attributes to the **product entity**. But in my case I want to add extension attributes to the **product attribute entity**. It's a different kind of entity. And at the moment there does not seem to be a way to do this in M2.2.4.

However, I see that the `getExtensionAttributes()`-method [is added to the 2.3-develop branch](https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Catalog/Api/Data/ProductAttributeInterface.php). Is there any chance this method will also be applied to the 2.2-develop branch so that it can be implemented in 2.2.5+?

I'm currently using an anonymous class to generate the class that would otherwise have been automatically generated:

    if ($attribute->getExtensionAttributes() === null) {
        // Temporary solution because ProductAttributeInterface does not seem to support extension attributes yet.
        $extensionAttributes = new class implements AttributeExtensionInterface
        {
            /**
             * @var string
             */
            protected $attributeSuffix;

            /**
             * @var boolean
             */
            protected $trimTrailingZeros;

            /**
             * @return string
             */
            public function getAttributeSuffix(): string
            {
                return $this->attributeSuffix;
            }

            /**
             * @param string $attributeSuffix
             */
            public function setAttributeSuffix(string $attributeSuffix): void
            {
                $this->attributeSuffix = $attributeSuffix;
            }

            /**
             * @return bool
             */
            public function getTrimTrailingZeros(): bool
            {
                return $this->trimTrailingZeros;
            }

            /**
             * @param mixed $trimTrailingZeros
             */
            public function setTrimTrailingZeros(bool $trimTrailingZeros): void
            {
                $this->trimTrailingZeros = $trimTrailingZeros;
            }
        };
        $attribute->setExtensionAttributes($extensionAttributes);
    }

**Comment by ghost** [comment]
Hi @kanduvisla The fix for this issue is already available in the 2.3-develop branch.
Thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by kanduvisla** [comment]
For anyone who comes across this thread:

Don't use anonymous class in Magento 2.2: there is a known issue in `setup:di:compile` that will throw an error when you're using anonymous classes.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by ghost** [comment]
Hi @kanduvisla this issue still present  on 2.3-develop branch
![selection_206](https://user-images.githubusercontent.com/40756439/48554169-53bfc900-e8e6-11e8-8882-ca79a30fba4f.png)

**Comment by tylerssn** [comment]
@kanduvisla - do you have a record of the solution you put in place to get around this in 2.2?

**Comment by drc0** [comment]
Hello, not sure if that has been fixed in 2.3.4 but I cannot add extension_attributes on ProductAttributeInterface, 


```
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Api/etc/extension_attributes.xsd">
    <extension_attributes for="Magento\Catalog\Api\Data\ProductAttributeInterface">
        <attribute code="option_guids" type="V\M\Api\Data\ProductAttributeGuidOptionInterface[]" />
    </extension_attributes>
</config>
```
```
$e = $attribute->getExtensionAttributes();
$e->getOptionGuids(); // PHP Error:  Call to undefined method Magento/Eav/Api/Data/AttributeExtension::getOptionGuids()

```

is there a patch for this?
